node/src/chain/fallback.rs

```diff
@@ -33,67 +36,86 @@ impl<'a, N: Network, DB: database::DB, VM: vm::VMExecution>
+            (_, Ordering::Greater) => Err(anyhow!(
+                "iteration {:?} is higher than the current {:?}",
+                remote.iteration,
+                local.iteration
+            )),
+            (_, Ordering::Equal) => Err(anyhow!(
+                "iteration is equal to the current {:?}",
+                local.iteration
+            )),
+            _ => Ok(()),
+        }?;
+
+        let (prev_header, prev_prev_header) =
+            self.acc.db.read().await.view(|t| {
+                let (prev_block_header, _) = t
+                    .fetch_block_header(&local.prev_block_hash)?
+                    .expect("block must exist");
+
+                let (prev_prev_block_header, _) = t
+                    .fetch_block_header(&prev_block_header.prev_block_hash)?
+                    .expect("block must exist");
+
+                Ok::<(ledger::Header, ledger::Header), anyhow::Error>((
+                    prev_block_header,
+                    prev_prev_block_header,
+                ))
+            })?;
 
         info!(
-            event = "starting fallback",
-            height = curr_height,
-            iter = curr_iteration,
-            target_iter = blk.header().iteration,
+            event = "execute fallback checks",
+            height = local.height,
+            iter = local.iteration,
+            target_iter = remote.iteration,
         );
 
-        let prev_block_height = curr_height - 1;
-        let prev_block = acc.db.read().await.view(|v| {
-            Ledger::fetch_block_by_height(&v, prev_block_height)?
-                .ok_or_else(|| anyhow::anyhow!("could not fetch block"))
-        })?;
-
-        info!(
-            event = "fallback checking block",
-            height = curr_height,
-            iter = curr_iteration,
-            target_iter = blk.header().iteration,
-        );
-
-        // Validate Header/Certificate of the new block upon previous block and
-        // provisioners.
-
-        // In an edge case, this may fail on performing fallback between two
-        // epochs.
-        let provisioners_list = acc.provisioners_list.read().await;
-        acceptor::verify_block_header(
+        let provisioners_list = self
+            .acc
+            .vm
+            .read()
+            .await
+            .get_provisioners(prev_header.state_hash)?;
+
+        let prev_provisioners_list = self
+            .acc
+            .vm
+            .read()
+            .await
+            .get_provisioners(prev_prev_header.state_hash)?;
+
+        let mut provisioners_list = ContextProvisioners::new(provisioners_list);
+        provisioners_list.set_previous(prev_provisioners_list);
+
+        // Ensure header of the new block is valid according to prev_block
+        // header
+        let _ = acceptor::verify_block_header(
             self.acc.db.clone(),
-            prev_block.header(),
+            &prev_header,
```

Comment:
I think we should rename the `mrb` argument in `verify_block_header` to `prev_block`